src/coreclr/src/jit/flowgraph.cpp

```diff
@@ -3557,10 +3557,165 @@ void Compiler::fgInitBlockVarSets()
+            continue;
+        }
+
+        // This block needs a GC poll. We either just insert a callout or we split the block and inline part of
+        // the test.
+
+        // If we're doing GCPOLL_CALL, just insert a GT_CALL node before the last node in the block.
+        CLANG_FORMAT_COMMENT_ANCHOR;
+
+#ifdef DEBUG
+        switch (block->bbJumpKind)
+        {
+            case BBJ_RETURN:
+            case BBJ_ALWAYS:
+            case BBJ_COND:
+            case BBJ_SWITCH:
+            case BBJ_NONE:
+            case BBJ_THROW:
+                break;
+            default:
+                assert(!"Unexpected block kind");
+        }
+#endif // DEBUG
+
+        GCPollType pollType = GCPOLL_INLINE;
+
+        // We'd like to inset an inline poll. Below is the list of places where we
+        // can't or don't want to emit an inline poll. Check all of those. If after all of that we still
+        // have INLINE, then emit an inline check.
+
+        if (opts.OptimizationDisabled())
+        {
+#ifdef DEBUG
+            if (verbose)
+            {
+                printf("Selecting CALL poll in block " FMT_BB " because of debug/minopts\n", block->bbNum);
+            }
+#endif // DEBUG
+
+            // Don't split blocks and create inlined polls unless we're optimizing.
+            pollType = GCPOLL_CALL;
+        }
+        else if (genReturnBB == block)
+        {
+#ifdef DEBUG
+            if (verbose)
+            {
+                printf("Selecting CALL poll in block " FMT_BB " because it is the single return block\n", block->bbNum);
+            }
+#endif // DEBUG
+
+            // we don't want to split the single return block
+            pollType = GCPOLL_CALL;
+        }
+        else if (BBJ_SWITCH == block->bbJumpKind)
+        {
+#ifdef DEBUG
+            if (verbose)
+            {
+                printf("Selecting CALL poll in block " FMT_BB " because it is a SWITCH block\n", block->bbNum);
+            }
+#endif // DEBUG
+
+            // We don't want to deal with all the outgoing edges of a switch block.
+            pollType = GCPOLL_CALL;
+        }
+
+        BasicBlock* curBasicBlock = fgCreateGCPoll(pollType, block);
+        createdPollBlocks |= (block != curBasicBlock);
+        block = curBasicBlock;
+    }
+
+    // If we split a block to create a GC Poll, then rerun fgReorderBlocks to push the rarely run blocks out
+    // past the epilog.  We should never split blocks unless we're optimizing.
+    if (createdPollBlocks)
+    {
+        noway_assert(opts.OptimizationEnabled());
+        fgReorderBlocks();
+        fgUpdateChangedFlowGraph();
+    }
+#ifdef DEBUG
+    if (verbose)
+    {
+        printf("*************** After fgInsertGCPolls()\n");
+        fgDispBasicBlocks(true);
+    }
+#endif // DEBUG
+}
+
 /*****************************************************************************
  *
  *  The following does the final pass on BBF_NEEDS_GCPOLL and then actually creates the GC Polls.
  */
+
```

Comment:
Add new-style header comment?

Reply:
I will remove this method in a subsequent PR.